File: src/IR.cpp
//===- IR.cpp ---------------------------------------------------*- C++ -*-===//
//
//  Copyright (C) 2020 GrammaTech, Inc.
//
//  This code is licensed under the MIT license. See the LICENSE file in the
//  project root for license terms.
//
//  This project is sponsored by the Office of Naval Research, One Liberty
//  Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
//  N68335-17-C-0700.  The content of the information does not necessarily
//  reflect the position or policy of the Government and no official
//  endorsement should be inferred.
//
//===----------------------------------------------------------------------===//
#include "CFGSerialization.hpp"
#include "Serialization.hpp"
#include <gtirb/DataBlock.hpp>
#include <gtirb/IR.hpp>
#include <gtirb/Module.hpp>
#include <gtirb/Section.hpp>
#include <gtirb/Symbol.hpp>
#include <gtirb/SymbolicExpression.hpp>
#include <gtirb/proto/IR.pb.h>
#include <google/protobuf/io/coded_stream.h>
#include <google/protobuf/io/zero_copy_stream_impl.h>
#include <google/protobuf/util/json_util.h>
#include <iostream>
#include <memory>

using namespace gtirb;

static constexpr const char* GTIRB_MAGIC_CHARS = "GTIRB";

class IR::ModuleObserverImpl : public ModuleObserver {
public:
  explicit ModuleObserverImpl(IR* I_) : I(I_) {}

  ChangeStatus nameChange(Module* M, const std::string& /*OldName*/,
                          const std::string& /*NewName*/) override {
    auto& Index = I->Modules.get<by_pointer>();
    auto It = Index.find(M);
    assert(It != Index.end() && "module observed by non-owner");
    // The lambda would ordinarily update the Module such that the result
    // of Module::getName changes. Because that change happened before this
    // method was called, the lambda doesn't need to do anything.
    Index.modify(It, [](Module*) {});
    return ChangeStatus::Accepted;
  }

  ChangeStatus addProxyBlocks(Module* /*M*/,
                              Module::proxy_block_range Blocks) override {
    ChangeStatus Status = ChangeStatus::NoChange;
    if (!Blocks.empty()) {
      for (ProxyBlock& PB : Blocks) {
        // User could have called addVertex themselves, so check whether we
        // actually modified the graph.
        if (addVertex(&PB, I->Cfg).second)
          Status = ChangeStatus::Accepted;
      }
    }
    return Status;
  }

  ChangeStatus removeProxyBlocks(Module* /*M*/,
                                 Module::proxy_block_range Blocks) override {
    ChangeStatus Status = ChangeStatus::NoChange;
    if (!Blocks.empty()) {
      for (ProxyBlock& PB : Blocks) {
        // User could have called removeVertex themselves, so check whether
        // we actually modified the graph.
        if (removeVertex(&PB, I->Cfg))
          Status = ChangeStatus::Accepted;
      }
    }
    return Status;
  }

  ChangeStatus addCodeBlocks(Module* /*M*/,
                             Module::code_block_range Blocks) override {
    ChangeStatus Status = ChangeStatus::NoChange;
    if (!Blocks.empty()) {
      for (CodeBlock& CB : Blocks) {
        // User could have called addVertex themselves, so check whether we
        // actually modified the graph.
        if (addVertex(&CB, I->Cfg).second)
          Status = ChangeStatus::Accepted;
      }
    }
    return Status;
  }

  ChangeStatus removeCodeBlocks(Module* /*M*/,
                                Module::code_block_range Blocks) override {
    ChangeStatus Status = ChangeStatus::NoChange;
    if (!Blocks.empty()) {
      for (CodeBlock& CB : Blocks) {
        // User could have called removeVertex themselves, so check whether
        // we actually modified the graph.
        if (removeVertex(&CB, I->Cfg))
          Status = ChangeStatus::Accepted;
      }
    }
    return Status;
  }

private:
  IR* I;
};

IR::IR(Context& C)
    : AuxDataContainer(C, Kind::IR),
      MO(std::make_unique<ModuleObserverImpl>(this)) {}

IR::IR(Context& C, const UUID& U)
    : AuxDataContainer(C, Kind::IR, U),
      MO(std::make_unique<ModuleObserverImpl>(this)) {}

class IRLoadErrorCategory : public std::error_category {
public:
  [[nodiscard]] const char* name() const noexcept override {
    return "gt.gtirb.ir";
  }
  [[nodiscard]] std::string message(int Condition) const override {
    switch (static_cast<IR::load_error>(Condition)) {
    case IR::load_error::IncorrectVersion:
      return "Incompatible protobuf version";
    case IR::load_error::CorruptFile:
      return "Corrupted GTIRB file";
    case IR::load_error::CorruptModule:
      return "Corrupted GTIRB module";
    case IR::load_error::CorruptSection:
      return "Corrupted GTIRB section";
    case IR::load_error::CorruptByteInterval:
      return "Corrupted byte interval";
    case IR::load_error::CorruptCFG:
      return "Error in parsing CFG";
    case IR::load_error::BadUUID:
      return "Bytes not valid UUID";
    case IR::load_error::MissingUUID:
      return "Could not locate UUID";
    case IR::load_error::NotGTIRB:
      return "File does not contain GTIRB";
    }
    assert(false && "Expected to handle all error codes");
    return "";
  }
};

const std::error_category& gtirb::loadErrorCategory() {
  static IRLoadErrorCategory Cat;
  return Cat;
}

void IR::toProtobuf(MessageType* Message) const {
  nodeUUIDToBytes(this, *Message->mutable_uuid());
  *Message->mutable_cfg() = gtirb::toProtobuf(this->Cfg);
  containerToProtobuf(this->Modules, Message->mutable_modules());
  AuxDataContainer::toProtobuf(Message);
  Message->set_version(Version);
}

ErrorOr<IR*> IR::fromProtobuf(Context& C, const MessageType& Message) {
  UUID Id;
  if (!uuidFromBytes(Message.uuid(), Id))
    return {load_error::CorruptFile, "Cannot load IR"};

  auto* I = IR::Create(C, Id);
  int i = 0;
  for (const auto& Elt : Message.modules()) {
    auto M = Module::fromProtobuf(C, Elt);
    if (!M) {
      ErrorInfo Err{load_error::CorruptModule, "#" + std::to_string(i)};
      Err.Msg += "\n" + M.getError().message();
      return Err;
    }
    I->addModule(*M);
    ++i;
  }
  if (!gtirb::fromProtobuf(C, I->Cfg, Message.cfg()))
    return load_error::CorruptCFG;
  static_cast<AuxDataContainer*>(I)->fromProtobuf(Message);
  I->Version = Message.version();

  if (I->Version != GTIRB_PROTOBUF_VERSION) {
    std::stringstream ss;
    ss << I->Version << "; expected version " << GTIRB_PROTOBUF_VERSION;
    return {load_error::IncorrectVersion, ss.str()};
  }
  return I;
}

void IR::save(std::ostream& Out) const {
  // Magic signature
  // Magic signature
  // Bytes 0-4 contain the ASCII characters: GTIRB.
  // Bytes 5-6 are considered reserved for future use and should be 0.
  // Byte 7 contains the GTIRB protobuf spec version in use.
  Out << GTIRB_MAGIC_CHARS << static_cast<uint8_t>(0) << static_cast<uint8_t>(0)
      << static_cast<uint8_t>(GTIRB_PROTOBUF_VERSION);

  // Protobuf
  MessageType Message;
  this->toProtobuf(&Message);
  Message.SerializeToOstream(&Out);
}

ErrorOr<IR*> IR::load(Context& C, std::istream& In) {
  constexpr size_t magic_len =
      std::string::traits_type::length(GTIRB_MAGIC_CHARS);
  std::array<char, magic_len> magic;
  In.read(magic.data(), magic_len);
  if (memcmp(magic.data(), GTIRB_MAGIC_CHARS, magic_len) != 0) {
    return {load_error::NotGTIRB, "GTIRB magic signature not found"};
  }

  uint8_t res0;
  In >> res0;

  uint8_t res1;
  In >> res1;

  uint8_t protobuf_version;
  In >> protobuf_version;
  if (protobuf_version != GTIRB_PROTOBUF_VERSION) {
    std::stringstream ss;
    ss << "GTIRB protobuf version mismatch. Expected: "
       << GTIRB_PROTOBUF_VERSION << " Saw: " << protobuf_version;
    return {load_error::IncorrectVersion, ss.str()};
  }

  google::protobuf::io::IstreamInputStream InputStream(&In);
  google::protobuf::io::CodedInputStream CodedStream(&InputStream);
#ifdef PROTOBUF_SET_BYTES_LIMIT
  CodedStream.SetTotalBytesLimit(INT_MAX, INT_MAX);
#endif

  MessageType Message;
  if (!Message.ParseFromCodedStream(&CodedStream)) {
    return {load_error::CorruptFile, "Protobuf unable to be parsed"};
  }

  return IR::fromProtobuf(C, Message);
}

void IR::saveJSON(std::ostream& Out) const {
  MessageType Message;
  this->toProtobuf(&Message);
  std::string S;
  auto status = google::protobuf::util::MessageToJsonString(Message, &S);
  if (!status.ok()) {
    Out << status;
    return;
  } else {
    Out << S;
  }
}

ErrorOr<IR*> IR::loadJSON(Context& C, std::istream& In) {
  MessageType Message;
  std::string S;
  auto status = google::protobuf::util::JsonStringToMessage(
      std::string(std::istreambuf_iterator<char>(In), {}), &Message);
  if (!status.ok()) {
    return {load_error::CorruptFile, status.ToString()};
  } else {
    return IR::fromProtobuf(C, Message);
  }
}

File: java/com/grammatech/gtirb/IR.java
/*
 *  Copyright (C) 2020-2023 GrammaTech, Inc.
 *
 *  This code is licensed under the MIT license. See the LICENSE file in the
 *  project root for license terms.
 *
 *  This project is sponsored by the Office of Naval Research, One Liberty
 *  Center, 875 N. Randolph Street, Arlington, VA 22203 under contract #
 *  N68335-17-C-0700.  The content of the information does not necessarily
 *  reflect the position or policy of the Government and no official
 *  endorsement should be inferred.
 *
 */

package com.grammatech.gtirb;

import com.grammatech.gtirb.Module;
import com.grammatech.gtirb.proto.IROuterClass;
import com.grammatech.gtirb.proto.ModuleOuterClass;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * A complete internal representation. IR describes the internal representation
 * of a software artifact.
 */
public class IR extends AuxDataContainer {

    private List<Module> modules;
    private CFG cfg;
    private int version; // This is the protobuf version from the protoIr

    static final private byte[] GTIRB_MAGIC_CHARS =
        "GTIRB".getBytes(Charset.forName("ASCII"));
    static final private int GTIRB_MAGIC_LENGTH = 5;

    /**
     * Default class constructor for IR.
     */
    public IR() {
        super();
        this.modules = new ArrayList<Module>();
    }

    /**
     * Class constructor for IR from an IR protobuf.
     * @param  protoIr  The {@link IR} as serialized into a protocol buffer.
     */
    private IR(IROuterClass.IR protoIr) {
        super(protoIr.getUuid(), protoIr.getAuxDataMap());
    }

    /**
     * Load IR from protobuf.
     *
     * @return  The {@link IR} loaded from the protobuf.
     */
    private static IR loadProtobuf(IROuterClass.IR protoIr) {
        // If no protobuf, can't load it.
        if (protoIr == null)
            return null;
        IR ir = new IR(protoIr);
        ir.version = protoIr.getVersion();
        // Import the modules
        ir.modules = new ArrayList<Module>();
        for (ModuleOuterClass.Module protoModule : protoIr.getModulesList()) {
            Module module = Module.fromProtobuf(protoModule);
            ir.modules.add(module);
        }
        // Import the CFG
        ir.cfg = new CFG(protoIr.getCfg());
        return ir;
    }

    /**
     * Load IR from a protobuf file stream.
     *
     * @return  IR if load is successful, null otherwise.
     */
    public static IR loadFile(InputStream fileIn) {
        byte[] magic = new byte[GTIRB_MAGIC_LENGTH];
        try {
            // Magic signature
            // Bytes 0-4 contain the ASCII characters: GTIRB.
            // Bytes 5-6 are considered reserved for future use and should be 0.
            // Byte 7 contains the GTIRB protobuf spec version in use.
            int bytes_read = fileIn.read(magic);
            if (bytes_read != GTIRB_MAGIC_LENGTH ||
                !Arrays.equals(magic, GTIRB_MAGIC_CHARS)) {
                return null;
            }
            fileIn.skip(2);
            int ver = fileIn.read();
            if (ver != Version.gtirbProtobufVersion) {
                return null;
            }
        } catch (IOException ie) {
            return null;
        }

        IROuterClass.IR protoIr;
        try {
            protoIr = IROuterClass.IR.parseFrom(fileIn);
        } catch (FileNotFoundException fe) {
            return null;
        } catch (IOException ie) {
            return null;
        }
        IR ir = IR.loadProtobuf(protoIr);
        return ir;
    }

    /**
     * Load IR from a protobuf file.
     *
     * @return  IR if load is successful, null otherwise.
     */
    public static IR loadFile(String fileInName) {
        try {
            File fileIn = new File(fileInName);
            FileInputStream fileInputStream = new FileInputStream(fileIn);
            return loadFile(fileInputStream);
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Get the list of modules belonging to this {@link IR}.
     *
     * @return  An unmodifiable {@link Module} list of all the
     * modules in this {@link IR}. Any attempt to remove an element of
     * this list will throw an UnsupportedOperationException.
     */
    public List<Module> getModules() {
        return Collections.unmodifiableList(this.modules);
    }

    /**
     * Add a module to this {@link IR}.
     *
     * @param module  {@link Module} to add.
     */
    public void addModule(Module module) {
        this.modules.add(module);
        module.setIr(Optional.of(this));
    }

    /**
     * Add a list of modules to this {@link IR}.
     *
     * @param modules  Modules to add.
     */
    public void addModules(List<Module> modules) {
        for (Module module : modules) {
            this.addModule(module);
        }
    }

    /**
     * Remove a module from this {@link IR}.
     *
     * @param module  {@link Module} to remove.
     */
    public void removeModule(Module module) {
        this.modules.remove(module);
        module.setIr(Optional.empty());
    }

    /**
     * Get the CFG belonging to this {@link IR}.
     *
     * @return  A {@link CFG}.
     */
    public CFG getCfg() { return this.cfg; }

    /**
     * Set the CFG belonging to this {@link IR}.
     *
     * @param cfg  A {@link CFG}.
     */
    public void setCfg(CFG cfg) { this.cfg = cfg; }

    /**
     * Get the protobuf version of this {@link IR}.
     *
     * @return Protobuf version.
     */
    public int getVersion() { return this.version; }

    /**
     * Set the protobuf version of this {@link IR}.
     *
     * @param version Protobuf version.
     */
    public void setVersion(int version) { this.version = version; }

    /**
     * Serialize this IR into a protobuf.
     *
     * @return IR protocol buffer.
     */
    public IROuterClass.IR.Builder toProtobuf() {
        IROuterClass.IR.Builder protoIr = IROuterClass.IR.newBuilder();
        protoIr.setUuid(Util.uuidToByteString(this.getUuid()));
        protoIr.setVersion(Version.gtirbProtobufVersion);
        // Add modules
        for (Module module : this.modules) {
            ModuleOuterClass.Module.Builder protoModule = module.toProtobuf();
            protoIr.addModules(protoModule);
        }
        // Add CFG
        if (this.cfg != null) {
            protoIr.setCfg(this.cfg.toProtobuf());
        }
        return protoIr;
    }

    /**
     * Save IR to a protobuf file stream.
     */
    public void saveFile(OutputStream fileOut) throws IOException {
        fileOut.write(GTIRB_MAGIC_CHARS);
        fileOut.write(0);
        fileOut.write(0);
        fileOut.write(Version.gtirbProtobufVersion);

        IROuterClass.IR protoIr = this.toProtobuf().build();
        protoIr.writeTo(fileOut);
    }

    /**
     * Save IR to a protobuf file.
     */
    public void saveFile(String fileOutName) throws IOException {
        File fileOut = new File(fileOutName);
        FileOutputStream fileOutputStream = new FileOutputStream(fileOut);
        this.saveFile(fileOutputStream);
    }
}
